### ISSUE TITLE
[Bug] Yarn instal fails if node_modules is a docker volume or similar

### ISSUE THREAD

**Issue by ncjones** [yarnpkg/berry#4172]
This is the exact same issue as #1184.

When using nodeLinker: node-modules, during the link step, yarn will attempt to delete the entire node_modules directory, instead of only its contents. This causes problems with the common docker pattern of mounting node_modules as a volume, resulting in an error like the following:

```
➤ YN0000: ┌ Link step
➤ YN0001: │ Error: EBUSY: resource busy or locked, rmdir '/work/node_modules'
➤ YN0000: └ Completed
➤ YN0000: Failed with errors in 0s 98ms
```

### To reproduce

```
git clone git@github.com:ncjones/yarn-nm-volume.git
cd yarn-nm-volume
docker compose run script yarn install
```

See [ncjones/yarn-nm-volume](https://github.com/ncjones/yarn-nm-volume).

### Environment

```shell
System:
    OS: macOS
```

**Comment by OACDesigns** [comment]
Also having this problem since upgrading to 3.2 (I had to upgrade because of a different bug in an older version resulting in npm packages not resolving).  

Have you found a workaround?  
Or perhaps a previous version before this behaviour was introduced?

**Comment by larixer** [comment]
I have opened PR to fix this problem: #4206
To try the fix from PR, until it is merged and released, run:
`yarn set version from sources --branch 4206`
